cirq/ops/probable_gate.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numbers
+from typing import Tuple, TYPE_CHECKING, Dict, Any
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality(approximate=True)
+class ProbableGate(raw_types.Gate):
+    """Applies a sub operation with some probability."""
+
+    def __init__(self, *, sub_gate: 'cirq.Gate', probability: value.TParamVal):
+        if isinstance(probability,
+                      numbers.Number) and not 0 <= float(probability) <= 1:
+            raise ValueError("not 0 <= probability <= 1")
+
+        self.sub_gate = sub_gate
+        self.probability = probability
+
+        # Auto flatten.
+        if isinstance(self.sub_gate, ProbableGate):
+            self.probability *= self.sub_gate.probability
+            self.sub_gate = self.sub_gate.sub_gate
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return protocols.qid_shape(self.sub_gate)
+
+    def _value_equality_values_(self):
+        return self.sub_gate, self.probability
+
+    def _has_unitary_(self):
+        return False
```

Comment:
what about probability=0 and gate is unitary?

Reply:
Which it's technically true that when probability is 0 or 1 that there is a unitary effect, I think it's fair to treat these cases uniformly with the rest. Similar to how the bit flip channel isn't considered unitary when p=1 and the various damping channels aren't unitary when gamma=0.

Reply:
SGTM